File: tl/src/main/java/com/github/badoualy/telegram/tl/api/TLChannel.java
package com.github.badoualy.telegram.tl.api;

import com.github.badoualy.telegram.tl.TLContext;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readLong;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLObject;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeLong;
import static com.github.badoualy.telegram.tl.StreamUtils.writeString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLObject;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLChannel extends TLAbsChat {

    public static final int CONSTRUCTOR_ID = 0xa14dca52;

    protected int flags;

    protected boolean creator;

    protected boolean kicked;

    protected boolean left;

    protected boolean editor;

    protected boolean moderator;

    protected boolean broadcast;

    protected boolean verified;

    protected boolean megagroup;

    protected boolean restricted;

    protected boolean democracy;

    protected boolean signatures;

    protected boolean min;

    protected Long accessHash;

    protected String title;

    protected String username;

    protected TLAbsChatPhoto photo;

    protected int date;

    protected int version;

    protected String restrictionReason;

    private final String _constructor = "channel#a14dca52";

    public TLChannel() {
    }

    public TLChannel(boolean creator, boolean kicked, boolean left, boolean editor, boolean moderator, boolean broadcast, boolean verified, boolean megagroup, boolean restricted, boolean democracy, boolean signatures, boolean min, int id, Long accessHash, String title, String username, TLAbsChatPhoto photo, int date, int version, String restrictionReason) {
        this.creator = creator;
        this.kicked = kicked;
        this.left = left;
        this.editor = editor;
        this.moderator = moderator;
        this.broadcast = broadcast;
        this.verified = verified;
        this.megagroup = megagroup;
        this.restricted = restricted;
        this.democracy = democracy;
        this.signatures = signatures;
        this.min = min;
        this.id = id;
        this.accessHash = accessHash;
        this.title = title;
        this.username = username;
        this.photo = photo;
        this.date = date;
        this.version = version;
        this.restrictionReason = restrictionReason;
    }

    private void computeFlags() {
        flags = 0;
        flags = creator ? (flags | 1) : (flags & ~1);
        flags = kicked ? (flags | 2) : (flags & ~2);
        flags = left ? (flags | 4) : (flags & ~4);
        flags = editor ? (flags | 8) : (flags & ~8);
        flags = moderator ? (flags | 16) : (flags & ~16);
        flags = broadcast ? (flags | 32) : (flags & ~32);
        flags = verified ? (flags | 128) : (flags & ~128);
        flags = megagroup ? (flags | 256) : (flags & ~256);
        flags = restricted ? (flags | 512) : (flags & ~512);
        flags = democracy ? (flags | 1024) : (flags & ~1024);
        flags = signatures ? (flags | 2048) : (flags & ~2048);
        flags = min ? (flags | 4096) : (flags & ~4096);
        flags = accessHash != null ? (flags | 8192) : (flags & ~8192);
        flags = username != null ? (flags | 64) : (flags & ~64);
        flags = restrictionReason != null ? (flags | 512) : (flags & ~512);

        // Following parameters might be forced to true by another field that updated the flags
        restricted = (flags & 512) != 0;
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        computeFlags();

        writeInt(flags, stream);
        writeInt(id, stream);
        if ((flags & 8192) != 0) {
            if (accessHash == null) throwNullFieldException("accessHash", flags);
            writeLong(accessHash, stream);
        }
        writeString(title, stream);
        if ((flags & 64) != 0) {
            if (username == null) throwNullFieldException("username", flags);
            writeString(username, stream);
        }
        writeTLObject(photo, stream);
        writeInt(date, stream);
        writeInt(version, stream);
        if ((flags & 512) != 0) {
            if (restrictionReason == null) throwNullFieldException("restrictionReason", flags);
            writeString(restrictionReason, stream);
        }
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        flags = readInt(stream);
        creator = (flags & 1) != 0;
        kicked = (flags & 2) != 0;
        left = (flags & 4) != 0;
        editor = (flags & 8) != 0;
        moderator = (flags & 16) != 0;
        broadcast = (flags & 32) != 0;
        verified = (flags & 128) != 0;
        megagroup = (flags & 256) != 0;
        restricted = (flags & 512) != 0;
        democracy = (flags & 1024) != 0;
        signatures = (flags & 2048) != 0;
        min = (flags & 4096) != 0;
        id = readInt(stream);
        accessHash = (flags & 8192) != 0 ? readLong(stream) : null;
        title = readTLString(stream);
        username = (flags & 64) != 0 ? readTLString(stream) : null;
        photo = readTLObject(stream, context, TLAbsChatPhoto.class, -1);
        date = readInt(stream);
        version = readInt(stream);
        restrictionReason = (flags & 512) != 0 ? readTLString(stream) : null;
    }

    @Override
    public int computeSerializedSize() {
        computeFlags();

        int size = SIZE_CONSTRUCTOR_ID;
        size += SIZE_INT32;
        size += SIZE_INT32;
        if ((flags & 8192) != 0) {
            if (accessHash == null) throwNullFieldException("accessHash", flags);
            size += SIZE_INT64;
        }
        size += computeTLStringSerializedSize(title);
        if ((flags & 64) != 0) {
            if (username == null) throwNullFieldException("username", flags);
            size += computeTLStringSerializedSize(username);
        }
        size += photo.computeSerializedSize();
        size += SIZE_INT32;
        size += SIZE_INT32;
        if ((flags & 512) != 0) {
            if (restrictionReason == null) throwNullFieldException("restrictionReason", flags);
            size += computeTLStringSerializedSize(restrictionReason);
        }
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public boolean getCreator() {
        return creator;
    }

    public void setCreator(boolean creator) {
        this.creator = creator;
    }

    public boolean getKicked() {
        return kicked;
    }

    public void setKicked(boolean kicked) {
        this.kicked = kicked;
    }

    public boolean getLeft() {
        return left;
    }

    public void setLeft(boolean left) {
        this.left = left;
    }

    public boolean getEditor() {
        return editor;
    }

    public void setEditor(boolean editor) {
        this.editor = editor;
    }

    public boolean getModerator() {
        return moderator;
    }

    public void setModerator(boolean moderator) {
        this.moderator = moderator;
    }

    public boolean getBroadcast() {
        return broadcast;
    }

    public void setBroadcast(boolean broadcast) {
        this.broadcast = broadcast;
    }

    public boolean getVerified() {
        return verified;
    }

    public void setVerified(boolean verified) {
        this.verified = verified;
    }

    public boolean getMegagroup() {
        return megagroup;
    }

    public void setMegagroup(boolean megagroup) {
        this.megagroup = megagroup;
    }

    public boolean getRestricted() {
        return restricted;
    }

    public void setRestricted(boolean restricted) {
        this.restricted = restricted;
    }

    public boolean getDemocracy() {
        return democracy;
    }

    public void setDemocracy(boolean democracy) {
        this.democracy = democracy;
    }

    public boolean getSignatures() {
        return signatures;
    }

    public void setSignatures(boolean signatures) {
        this.signatures = signatures;
    }

    public boolean getMin() {
        return min;
    }

    public void setMin(boolean min) {
        this.min = min;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public Long getAccessHash() {
        return accessHash;
    }

    public void setAccessHash(Long accessHash) {
        this.accessHash = accessHash;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public TLAbsChatPhoto getPhoto() {
        return photo;
    }

    public void setPhoto(TLAbsChatPhoto photo) {
        this.photo = photo;
    }

    public int getDate() {
        return date;
    }

    public void setDate(int date) {
        this.date = date;
    }

    public int getVersion() {
        return version;
    }

    public void setVersion(int version) {
        this.version = version;
    }

    public String getRestrictionReason() {
        return restrictionReason;
    }

    public void setRestrictionReason(String restrictionReason) {
        this.restrictionReason = restrictionReason;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/TLChannel.kt
package com.github.badoualy.telegram.tl.api

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException

/**
 * channel#cb44b1c
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLChannel() : TLAbsChat() {
    @Transient
    var creator: Boolean = false

    @Transient
    var left: Boolean = false

    @Transient
    var editor: Boolean = false

    @Transient
    var broadcast: Boolean = false

    @Transient
    var verified: Boolean = false

    @Transient
    var megagroup: Boolean = false

    @Transient
    var restricted: Boolean = false

    @Transient
    var democracy: Boolean = false

    @Transient
    var signatures: Boolean = false

    @Transient
    var min: Boolean = false

    override var id: Int = 0

    var accessHash: Long? = null

    var title: String = ""

    var username: String? = null

    var photo: TLAbsChatPhoto = TLChatPhotoEmpty()

    var date: Int = 0

    var version: Int = 0

    var restrictionReason: String? = null

    var adminRights: TLChannelAdminRights? = null

    var bannedRights: TLChannelBannedRights? = null

    private val _constructor: String = "channel#cb44b1c"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            creator: Boolean,
            left: Boolean,
            editor: Boolean,
            broadcast: Boolean,
            verified: Boolean,
            megagroup: Boolean,
            restricted: Boolean,
            democracy: Boolean,
            signatures: Boolean,
            min: Boolean,
            id: Int,
            accessHash: Long?,
            title: String,
            username: String?,
            photo: TLAbsChatPhoto,
            date: Int,
            version: Int,
            restrictionReason: String?,
            adminRights: TLChannelAdminRights?,
            bannedRights: TLChannelBannedRights?
    ) : this() {
        this.creator = creator
        this.left = left
        this.editor = editor
        this.broadcast = broadcast
        this.verified = verified
        this.megagroup = megagroup
        this.restricted = restricted
        this.democracy = democracy
        this.signatures = signatures
        this.min = min
        this.id = id
        this.accessHash = accessHash
        this.title = title
        this.username = username
        this.photo = photo
        this.date = date
        this.version = version
        this.restrictionReason = restrictionReason
        this.adminRights = adminRights
        this.bannedRights = bannedRights
    }

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(creator, 1)
        updateFlags(left, 4)
        updateFlags(editor, 8)
        updateFlags(broadcast, 32)
        updateFlags(verified, 128)
        updateFlags(megagroup, 256)
        updateFlags(restricted, 512)
        updateFlags(democracy, 1024)
        updateFlags(signatures, 2048)
        updateFlags(min, 4096)
        updateFlags(accessHash, 8192)
        updateFlags(username, 64)
        updateFlags(restrictionReason, 512)
        updateFlags(adminRights, 16384)
        updateFlags(bannedRights, 32768)

        // Following parameters might be forced to true by another field that updated the flags
        restricted = isMask(512)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        writeInt(id)
        doIfMask(accessHash, 8192) { writeLong(it) }
        writeString(title)
        doIfMask(username, 64) { writeString(it) }
        writeTLObject(photo)
        writeInt(date)
        writeInt(version)
        doIfMask(restrictionReason, 512) { writeString(it) }
        doIfMask(adminRights, 16384) { writeTLObject(it) }
        doIfMask(bannedRights, 32768) { writeTLObject(it) }
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        creator = isMask(1)
        left = isMask(4)
        editor = isMask(8)
        broadcast = isMask(32)
        verified = isMask(128)
        megagroup = isMask(256)
        restricted = isMask(512)
        democracy = isMask(1024)
        signatures = isMask(2048)
        min = isMask(4096)
        id = readInt()
        accessHash = readIfMask(8192) { readLong() }
        title = readString()
        username = readIfMask(64) { readString() }
        photo = readTLObject<TLAbsChatPhoto>()
        date = readInt()
        version = readInt()
        restrictionReason = readIfMask(512) { readString() }
        adminRights = readIfMask(16384) { readTLObject<TLChannelAdminRights>(TLChannelAdminRights::class, TLChannelAdminRights.CONSTRUCTOR_ID) }
        bannedRights = readIfMask(32768) { readTLObject<TLChannelBannedRights>(TLChannelBannedRights::class, TLChannelBannedRights.CONSTRUCTOR_ID) }
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += SIZE_INT32
        size += getIntIfMask(accessHash, 8192) { SIZE_INT64 }
        size += computeTLStringSerializedSize(title)
        size += getIntIfMask(username, 64) { computeTLStringSerializedSize(it) }
        size += photo.computeSerializedSize()
        size += SIZE_INT32
        size += SIZE_INT32
        size += getIntIfMask(restrictionReason, 512) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(adminRights, 16384) { it.computeSerializedSize() }
        size += getIntIfMask(bannedRights, 32768) { it.computeSerializedSize() }
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLChannel) return false
        if (other === this) return true

        return _flags == other._flags
                && creator == other.creator
                && left == other.left
                && editor == other.editor
                && broadcast == other.broadcast
                && verified == other.verified
                && megagroup == other.megagroup
                && restricted == other.restricted
                && democracy == other.democracy
                && signatures == other.signatures
                && min == other.min
                && id == other.id
                && accessHash == other.accessHash
                && title == other.title
                && username == other.username
                && photo == other.photo
                && date == other.date
                && version == other.version
                && restrictionReason == other.restrictionReason
                && adminRights == other.adminRights
                && bannedRights == other.bannedRights
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0xcb44b1c.toInt()
    }
}
